ComputeAdmin/AzureStack.ComputeAdmin.psm1

```diff
@@ -660,5 +662,185 @@ Function CreateGalleyItem{
+        $storageAccount = New-AzureRmStorageAccount -Name $storageAccountName -Location $location -ResourceGroupName $resourceGroupName -Type Standard_LRS
+    }
+
+    Set-AzureRmCurrentStorageAccount -StorageAccountName $storageAccountName -ResourceGroupName $resourceGroupName
+    
+    #same for container
+    $container = Get-AzureStorageContainer -Name $containerName -ErrorAction SilentlyContinue
+    if (-not ($container)) {
+        $container = New-AzureStorageContainer -Name $containerName -Permission Blob
+    }
+
+    if($pscmdlet.ParameterSetName -eq "VMExtensionFromLocal")
+    {
+        $storageAccount.PrimaryEndpoints.Blob
+        $script:extensionName = Split-Path $extensionLocalPath -Leaf
+        $script:extensionBlobURIFromLocal = '{0}{1}/{2}' -f $storageAccount.PrimaryEndpoints.Blob.AbsoluteUri, $containerName,$extensionName
+        Set-AzureStorageBlobContent -File $extensionLocalPath -Container $containerName -Blob $extensionName
+    }
+
+    $ArmEndpoint = $ArmEndpoint.TrimEnd("/")
+    $uri = $armEndpoint + '/subscriptions/' + $subscription + '/providers/Microsoft.Compute.Admin/locations/' + $location + '/artifactTypes/VMExtension/publishers/' + $publisher
+    $uri = $uri + '/types/' + $type + '/versions/' + $version + '?api-version=2015-12-01-preview'
+
+    #building request body JSON
+    if($pscmdlet.ParameterSetName -eq "VMExtensionFromLocal") {
+        $sourceBlobJSON = '"SourceBlob" : {"Uri" :"' + $extensionBlobURIFromLocal + '"}'
+    }
+    else {
+        $sourceBlobJSON = '"SourceBlob" : {"Uri" :"' + $extensionBlobURI + '"}'
+    }
+    
+    $osTypeJSON = '"VmOsType" : "' + $osType + '"'
+    $ComputeRoleJSON = '"ComputeRole" : "N/A"'
+    $VMScaleSetEnabledJSON = '"VMScaleSetEnabled" : "' + $vmScaleSetEnabled + '"'
+    $SupportMultipleExtensionsJSON = '"SupportMultipleExtensions" : "' + $supportMultipleExtensions + '"'
+    $IsSystemExtensionJSON = '"IsSystemExtension" : "false"'
+
+    $propertyBody = $sourceBlobJSON + "," + $osTypeJSON + ',' + $ComputeRoleJSON + "," + $VMScaleSetEnabledJSON + "," + $SupportMultipleExtensionsJSON + "," + $IsSystemExtensionJSON 
+
+    #building ARMResource
+
+    $RequestBody = '{"Properties":{'+$propertyBody+'}}'
+
+    Invoke-RestMethod -Method PUT -Uri $uri -Body $RequestBody -ContentType 'application/json' -Headers $Headers
+
+    $extensionHandler = Invoke-RestMethod -Method GET -Uri $uri -ContentType 'application/json' -Headers $Headers
+
+    while($extensionHandler.Properties.ProvisioningState -ne 'Succeeded')
+    {
+        if($extensionHandler.Properties.ProvisioningState -eq 'Failed')
+        {
+            Write-Error -Message "VM extension download failed." -ErrorAction Stop
+        }
+
+        if($extensionHandler.Properties.ProvisioningState -eq 'Canceled')
+        {
+            Write-Error -Message "VM extension download was canceled." -ErrorAction Stop
+        }
+
+        Write-Host "Downloading";
+        Start-Sleep -Seconds 4
```

Comment:
I see. Just curious because they use 10 seconds for waiting intervals. And personally I believe this waiting logic can be made a helper function instead. I'll open issue to improve this later.